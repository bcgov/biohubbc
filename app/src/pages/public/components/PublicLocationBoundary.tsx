import Box from '@material-ui/core/Box';
import Grid from '@material-ui/core/Grid';
import Typography from '@material-ui/core/Typography';
import InferredLocationDetails, { IInferredLayers } from 'components/boundary/InferredLocationDetails';
import MapContainer from 'components/map/MapContainer';
import { Feature } from 'geojson';
import { IGetProjectForViewResponse } from 'interfaces/useProjectApi.interface';
import React, { useEffect, useState } from 'react';
import { calculateUpdatedMapBounds } from 'utils/mapBoundaryUploadHelpers';

export interface IPublicLocationBoundaryProps {
  projectForViewData: IGetProjectForViewResponse;
  refresh: () => void;
}

/**
 * Location boundary content for a public (published) project.
 *
 * @return {*}
 */
const PublicLocationBoundary: React.FC<IPublicLocationBoundaryProps> = (props) => {
  const {
    projectForViewData: { location }
  } = props;

  const [inferredLayersInfo, setInferredLayersInfo] = useState<IInferredLayers>({
    parks: [],
    nrm: [],
    env: [],
    wmu: []
  });
  const [bounds, setBounds] = useState<any[] | undefined>([]);
  const [nonEditableGeometries, setNonEditableGeometries] = useState<any[]>([]);

  useEffect(() => {
    const nonEditableGeometriesResult = location.geometry.map((geom: Feature) => {
      return { feature: geom };
    });

    setBounds(calculateUpdatedMapBounds(location.geometry));
    setNonEditableGeometries(nonEditableGeometriesResult);
  }, [location.geometry]);

  return (
    <Box>
      <Box mb={2}>
        <Typography variant="h3">Project Location</Typography>
      </Box>
      <dl>
        <Grid container spacing={2}>
          <Grid item xs={12}>
            <Typography component="dt" variant="subtitle2" color="textSecondary">
              Location Description
            </Typography>
            <Typography component="dd" variant="body1">
              {location.location_description ? <>{location.location_description}</> : 'No Description'}
            </Typography>
          </Grid>
        </Grid>
      </dl>
      <Box mt={4} mb={4} height={500}>
        <MapContainer
          mapId="project_location_form_map"
          hideDrawControls={true}
          nonEditableGeometries={nonEditableGeometries}
          bounds={bounds}
          setInferredLayersInfo={setInferredLayersInfo}
        />
      </Box>
      <InferredLocationDetails layers={inferredLayersInfo} />
    </Box>
  );
};

export default PublicLocationBoundary;
