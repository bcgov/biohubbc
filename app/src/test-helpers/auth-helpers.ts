import { SYSTEM_ROLE } from 'constants/roles';
import { IAuthState } from 'contexts/authStateContext';
import Keycloak from 'keycloak-js';

/**
 * Represents an unauthenticated user who has:
 *  - not yet successfully authenticated (all keycloak details about the user will be false, null, or undefined)
 */
export const UnauthenticatedUserAuthState: IAuthState = {
  keycloakWrapper: {
    keycloak: {
      authenticated: false
    } as unknown as Keycloak,
    hasLoadedAllUserInfo: false,
    systemRoles: [],
    isSystemUser: () => false,
    hasSystemRole: () => false,
    hasAccessRequest: false,
    hasOneOrMoreProjectRoles: false,
    getUserIdentifier: () => null,
    getIdentitySource: () => null,
    getUserGuid: () => null,
    username: undefined,
    displayName: undefined,
    email: undefined,
    systemUserId: undefined,
    user: undefined,
    refresh: () => {
      // do nothing
    },
    getLoginUrl: () => '/login',
    critterbaseUuid: () => undefined
  }
};

/**
 * Represents an IDIR user who has:
 *  - successfully authenticated
 *  - has already been granted system access (has no pending access request)
 *  - has had all user info loaded successfully
 *  - has no system or project level roles
 */
export const SystemUserAuthState: IAuthState = {
  keycloakWrapper: {
    keycloak: {
      authenticated: true
    } as unknown as Keycloak,
    hasLoadedAllUserInfo: true,
    systemRoles: [],
    isSystemUser: () => true,
    hasSystemRole: () => false,
    hasAccessRequest: false,
    hasOneOrMoreProjectRoles: false,
    getUserIdentifier: () => 'testusername',
    getIdentitySource: () => 'idir',
    getUserGuid: () => '987-654-321',
    username: 'testusername',
    displayName: 'testdisplayname',
    email: 'test@email.com',
    systemUserId: 1,
    user: undefined,
    refresh: () => {
      // do nothing
    },
    getLoginUrl: () => '/login',
    critterbaseUuid: () => 'fakeguid'
  }
};

/**
 * Represents an IDIR user who has:
 *  - successfully authenticated
 *  - has already been granted system access (has no pending access request)
 *  - has had all user info loaded successfully
 *  - has the `System Administrator` system level role
 */
export const SystemAdminAuthState: IAuthState = {
  keycloakWrapper: {
    keycloak: {
      authenticated: true
    } as unknown as Keycloak,
    hasLoadedAllUserInfo: true,
    systemRoles: [SYSTEM_ROLE.SYSTEM_ADMIN],
    isSystemUser: () => true,
    hasSystemRole: () => true,
    hasAccessRequest: false,
    hasOneOrMoreProjectRoles: false,
    getUserIdentifier: () => 'admin-username',
    getIdentitySource: () => 'idir',
    getUserGuid: () => '123-456-789',
    username: 'admin-username',
    displayName: 'admin-displayname',
    email: 'admin@email.com',
    systemUserId: 1,
    user: undefined,
    refresh: () => {
      // do nothing
    },
    getLoginUrl: () => '/login',
    critterbaseUuid: () => 'fakeguid'
  }
};

// Same effect as `Partial` but applies to all levels of a nested object
type Subset<T> = {
  [P in keyof T]?: T[P] extends Record<any, any> | undefined ? Subset<T[P]> : T[P];
};

/**
 * Build and return a mock auth state object.
 *
 * @param {{ base: IAuthState; overrides?: Subset<IAuthState> }} options
 * @return {*}  {IAuthState}
 */
export const getMockAuthState = (options: { base: IAuthState; overrides?: Subset<IAuthState> }): IAuthState => {
  const { base, overrides } = options;

  return {
    ...base,
    ...overrides,
    keycloakWrapper: {
      ...base.keycloakWrapper,
      ...overrides?.keycloakWrapper,
      keycloak: {
        ...base.keycloakWrapper?.keycloak,
        ...overrides?.keycloakWrapper?.keycloak
      }
    }
  } as unknown as IAuthState;
};
