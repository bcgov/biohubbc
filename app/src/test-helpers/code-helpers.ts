import { IGetAllCodeSetsResponse } from 'interfaces/useCodesApi.interface';

export const codes: IGetAllCodeSetsResponse = {
  coordinator_agency: [{ id: 1, name: 'A Rocha Canada' }],
  management_action_type: [{ id: 1, name: 'Management action' }],
  first_nations: [{ id: 1, name: 'First nations code' }],
  funding_source: [{ id: 1, name: 'Funding source code' }],
  investment_action_category: [{ id: 1, fs_id: 1, name: 'Investment action category' }],
  activity: [{ id: 1, name: 'Activity code' }],
  project_type: [{ id: 1, name: 'Project type' }],
  region: [{ id: 1, name: 'Region code' }],
  species: [{ id: 1, name: 'Species code' }],
  proprietor_type: [
    { id: 1, name: 'Proprietor code 1', is_first_nation: false },
    { id: 2, name: 'First Nations Land', is_first_nation: true }
  ],
  iucn_conservation_action_level_1_classification: [{ id: 1, name: 'IUCN class' }],
  iucn_conservation_action_level_2_subclassification: [{ id: 1, iucn1_id: 1, name: 'IUCN subclass 1' }],
  iucn_conservation_action_level_3_subclassification: [{ id: 1, iucn2_id: 1, name: 'IUCN subclass 2' }],
  system_roles: [
    { id: 1, name: 'Role 1' },
    { id: 2, name: 'Role 2' }
  ],
  project_roles: [
    { id: 1, name: 'Project Role 1' },
    { id: 2, name: 'Project Role 2' }
  ],
  administrative_activity_status_type: [
    { id: 1, name: 'Pending' },
    { id: 2, name: 'Actioned' },
    { id: 3, name: 'Rejected' }
  ],
  common_survey_methodologies: [
    { id: 1, name: 'Recruitment' },
    { id: 2, name: 'SRB' }
  ]
};
