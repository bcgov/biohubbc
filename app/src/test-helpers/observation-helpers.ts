import { IGetObservationResponse } from 'interfaces/useObservationApi.interface';

export const getObservationForUpdateResponse: IGetObservationResponse = {
  id: 1,
  data: {
    metaData: {
      block_name: 1,
      block_size: 20,
      strata: 'Medium',
      date: '2020/04/04',
      start_time: '11:00',
      end_time: '11:00',
      pilot_name: 'Anissa',
      navigator: 'Agahchen',
      rear_left_observer: 'Shreyas',
      rear_right_observer: 'Deva',
      visibility: 'Moderate',
      light: 'Bright',
      cloud_cover: 20,
      temperature: 30,
      precipitation: 'Fog',
      wind_speed: 22,
      snow_cover: 12,
      snow_depth: 122,
      days_since_snowfall: 11,
      weather_description: 'Awesome weather',
      description_of_habitat: 'Habitat',
      aircraft_company: 'Boeing',
      aircraft_type: 'Jet',
      aircraft_registration_number: 123,
      aircraft_gps_model: 'Model',
      aircraft_gps_datum: 'NAD27',
      aircraft_gps_readout: 'UTM'
    },
    tableData: [[null, 1, 1, null, null, null, null, null, null, null, null, null, null, null]]
  }
};
