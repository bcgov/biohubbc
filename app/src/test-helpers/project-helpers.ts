import { IGetProjectForViewResponse } from 'interfaces/useProjectApi.interface';

export const getProjectForViewResponse: IGetProjectForViewResponse = {
  id: 1,
  project: {
    project_name: 'Test Project Name',
    project_type: '1',
    start_date: '1998-10-10',
    end_date: '2021-02-26',
    project_activities: [1],
    completion_status: 'Active',
    publish_date: '2021-01-26'
  },
  permit: {
    permits: [
      {
        permit_number: '123',
        permit_type: 'Permit type'
      }
    ]
  },
  location: {
    location_description: 'Location description',
    geometry: []
  },
  objectives: {
    objectives: 'Et ad et in culpa si',
    caveats: 'sjwer bds'
  },
  coordinator: {
    first_name: 'Amanda',
    last_name: 'Christensen',
    email_address: 'amanda@christensen.com',
    coordinator_agency: 'Amanda and associates',
    share_contact_details: 'true'
  },
  iucn: {
    classificationDetails: [
      {
        classification: 'classification',
        subClassification1: 'sub classification 1',
        subClassification2: 'sub classification 2'
      },
      {
        classification: 'classification again',
        subClassification1: 'sub classification 1 again',
        subClassification2: 'sub classification 2 again'
      }
    ]
  },
  funding: {
    fundingSources: [
      {
        id: 0,
        agency_id: 1,
        agency_name: 'agency name',
        agency_project_id: 'ABC123',
        investment_action_category: 1,
        investment_action_category_name: 'investment action',
        funding_amount: 333,
        start_date: '2000-04-14',
        end_date: '2021-04-13',
        revision_count: 1
      }
    ]
  },
  partnerships: {
    indigenous_partnerships: ['partner 1', 'partner 2'],
    stakeholder_partnerships: ['partner 3', 'partner 4']
  }
};
