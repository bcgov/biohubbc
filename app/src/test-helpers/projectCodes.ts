import { IGetAllCodesResponse } from 'interfaces/useBioHubApi-interfaces';

export const codes: IGetAllCodesResponse = {
  coordinator_agency: [],
  management_action_type: [],
  climate_change_initiative: [{ id: 1, name: 'climate code' }],
  first_nations: [],
  funding_source: [],
  investment_action_category: [],
  activity: [{ id: 1, name: 'activity code' }],
  project_type: [],
  region: [],
  species: [],
  iucn_conservation_action_level_1_classification: [],
  iucn_conservation_action_level_2_subclassification: [],
  iucn_conservation_action_level_3_subclassification: []
};
