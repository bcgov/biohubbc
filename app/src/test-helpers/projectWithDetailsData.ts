import { IProjectWithDetails } from 'interfaces/project-interfaces';

export const projectWithDetailsData: IProjectWithDetails = {
  id: 1,
  project: {
    project_name: 'Test Project Name',
    project_type_name: 'Type name',
    project_type: '1',
    start_date: '1998-10-10',
    end_date: '2021-02-26',
    climate_change_initiatives: [1],
    project_activities: [1]
  },
  location: {
    location_description: 'Location description',
    regions: ['Region 1', 'Region 2'],
    geometry: []
  },
  objectives: {
    objectives: 'Et ad et in culpa si',
    caveats: 'sjwer bds'
  },
  coordinator: {
    first_name: 'Amanda',
    last_name: 'Christensen',
    email_address: 'amanda@christensen.com',
    coordinator_agency: 'Amanda and associates',
    share_contact_details: 'true'
  }
};
