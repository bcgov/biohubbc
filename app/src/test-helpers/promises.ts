/**
 * A class used to wrap a generic, empty promise, which can be resolved or rejected
 * programmatically with a given value, thus "deferring" the promise.
 *
 * @example
 *
 * const deferred = new Deferred();
 *
 * const someDataFetcher = (callback: () => Promise<any>) => {
 *   callback().then((res) => {
 *     // res = 'some response data'
 *   });
 * };
 *
 * someDataFetcher(() => deferred.promise)
 *
 * deferred.resolve({ data: 'some response data' });
 *
 * @export
 * @class
 * @template R The type of the Promise.
 * @template Q The type of the reject reason.
 */
export class Deferred<R = void, Q extends any = void> {
  /**
   * Resolves the deferred promise with the given value.
   *
   * @param {R = void} value The value that the promise will resolve to.
   * @default
   * @returns {void}
   */
  resolve: (value: R) => void = () => null;

  /**
   * Rejects the deferred promise with the given reason.
   * @param {Q extends any = void} [reason] The reason that the promise rejects.
   * @returns {void}
   */
  reject: (reason?: Q) => void = () => null;

  /**
   * @property The promise being deferred.
   * @type {Promise<R>}
   */
  promise: Promise<R>;

  /**
   * Constructs a deferred promise.
   *
   * @constructor
   * @returns {Deffered} The deferred promise.
   */
  constructor() {
    this.promise = new Promise<R>((resolve, reject) => {
      this.resolve = resolve;
      this.reject = reject;
    });
  }

  /**
   * Resets the Deferred instance with a new unfulfilled promise.
   *
   * @returns {Deferred} The deferred promise.
   */
  reset() {
    this.promise = new Promise<R>((resolve, reject) => {
      this.resolve = resolve;
      this.reject = reject;
    });

    return this;
  }
}
