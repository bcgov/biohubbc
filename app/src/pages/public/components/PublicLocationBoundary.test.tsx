import { render, waitFor } from '@testing-library/react';
import React from 'react';
import { getProjectForViewResponse } from 'test-helpers/project-helpers';
import PublicLocationBoundary from './PublicLocationBoundary';

const mockRefresh = jest.fn();

jest.spyOn(console, 'debug').mockImplementation(() => {});

describe('PublicLocationBoundary', () => {
  it('renders correctly', async () => {
    const { asFragment } = render(
      <PublicLocationBoundary projectForViewData={getProjectForViewResponse} refresh={mockRefresh} />
    );

    await waitFor(() => {
      expect(asFragment()).toMatchSnapshot();
    });
  });
});
