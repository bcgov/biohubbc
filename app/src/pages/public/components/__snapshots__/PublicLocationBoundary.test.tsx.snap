// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PublicLocationBoundary renders correctly 1`] = `
<DocumentFragment>
  <div
    class="MuiBox-root MuiBox-root-1"
  >
    <div
      class="MuiBox-root MuiBox-root-2"
    >
      <h3
        class="MuiTypography-root MuiTypography-h3"
      >
        Project Location
      </h3>
    </div>
    <dl>
      <div
        class="MuiGrid-root MuiGrid-container MuiGrid-spacing-xs-2"
      >
        <div
          class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12"
        >
          <dt
            class="MuiTypography-root MuiTypography-subtitle2 MuiTypography-colorTextSecondary"
          >
            Location Description
          </dt>
          <dd
            class="MuiTypography-root MuiTypography-body1"
          >
            Location description
          </dd>
        </div>
      </div>
    </dl>
    <div
      class="MuiBox-root MuiBox-root-3"
    >
      <div
        class="leaflet-container leaflet-grab leaflet-touch-drag"
        id="project_location_form_map"
        style="height: 100%; position: relative;"
        tabindex="0"
      >
        <div
          class="leaflet-pane leaflet-map-pane"
          style="left: 0px; top: 0px;"
        >
          <div
            class="leaflet-pane leaflet-tile-pane"
          >
            <div
              class="leaflet-layer "
              style="z-index: 1;"
            >
              <div
                class="leaflet-tile-container leaflet-zoom-animated"
                style="z-index: 18; left: 0px; top: 0px;"
              >
                <img
                  alt=""
                  class="leaflet-tile"
                  role="presentation"
                  src="https://maps.gov.bc.ca/arcgis/rest/services/province/roads_wm/MapServer/tile/5/10/4"
                  style="width: 256px; height: 256px; left: -159px; top: -31px;"
                />
              </div>
            </div>
          </div>
          <div
            class="leaflet-pane leaflet-shadow-pane"
          />
          <div
            class="leaflet-pane leaflet-overlay-pane"
          />
          <div
            class="leaflet-pane leaflet-marker-pane"
          />
          <div
            class="leaflet-pane leaflet-tooltip-pane"
          />
          <div
            class="leaflet-pane leaflet-popup-pane"
          />
        </div>
        <div
          class="leaflet-control-container"
        >
          <div
            class="leaflet-top leaflet-left"
          >
            <div
              class="leaflet-control-zoom leaflet-bar leaflet-control"
            >
              <a
                aria-label="Zoom in"
                class="leaflet-control-zoom-in"
                href="#"
                role="button"
                title="Zoom in"
              >
                +
              </a>
              <a
                aria-label="Zoom out"
                class="leaflet-control-zoom-out"
                href="#"
                role="button"
                title="Zoom out"
              >
                âˆ’
              </a>
            </div>
            <div
              class="leaflet-control-fullscreen leaflet-bar leaflet-control"
            >
              <a
                class="leaflet-control-fullscreen-button leaflet-bar-part"
                href="#"
                title="View Fullscreen"
              />
            </div>
          </div>
          <div
            class="leaflet-top leaflet-right"
          >
            <div
              class="leaflet-draw leaflet-control"
            />
            <div
              class="leaflet-draw leaflet-control"
            />
          </div>
          <div
            class="leaflet-bottom leaflet-left"
          />
          <div
            class="leaflet-bottom leaflet-right"
          >
            <div
              aria-haspopup="true"
              class="leaflet-control-layers leaflet-control"
            >
              <a
                class="leaflet-control-layers-toggle"
                href="#"
                title="Layers"
              />
              <section
                class="leaflet-control-layers-list"
              >
                <div
                  class="leaflet-control-layers-base"
                >
                  <label>
                    <div>
                      <input
                        checked="checked"
                        class="leaflet-control-layers-selector"
                        name="leaflet-base-layers_34"
                        type="radio"
                      />
                      <span>
                         BC Government
                      </span>
                    </div>
                  </label>
                  <label>
                    <div>
                      <input
                        class="leaflet-control-layers-selector"
                        name="leaflet-base-layers_34"
                        type="radio"
                      />
                      <span>
                         Esri Imagery
                      </span>
                    </div>
                  </label>
                </div>
                <div
                  class="leaflet-control-layers-separator"
                  style="display: none;"
                />
                <div
                  class="leaflet-control-layers-overlays"
                />
              </section>
            </div>
            <div
              class="leaflet-control-attribution leaflet-control"
            >
              <a
                href="https://leafletjs.com"
                title="A JS library for interactive maps"
              >
                Leaflet
              </a>
            </div>
          </div>
        </div>
      </div>
    </div>
    <div
      class="MuiBox-root MuiBox-root-4"
    />
    <div
      class="MuiBox-root MuiBox-root-5"
    />
    <div
      class="MuiBox-root MuiBox-root-6"
    />
  </div>
</DocumentFragment>
`;
